lib/Staker/index.js

```diff
@@ -37,20 +38,20 @@ class Staker {
    * @param gatewayComposer Gateway composer contract address.
    */
   constructor(originWeb3, valueToken, brandedToken, gatewayComposer) {
-    const oThis = this;
-    oThis.originWeb3 = originWeb3;
-    oThis.valueToken = valueToken;
-    oThis.brandedToken = brandedToken;
-    oThis.gatewayComposer = gatewayComposer;
+    this.originWeb3 = originWeb3;
+    this.valueToken = valueToken;
+    this.brandedToken = brandedToken;
+    this.gatewayComposerAddress = gatewayComposer;
+
+    this.valueToken = new Mosaic.ContractInteract.EIP20Token(originWeb3, valueToken);
+    this.gatewayComposer = new GatewayComposer(originWeb3, gatewayComposer);
```

Comment:
Dependencies should be injected, not hard-coded.

Reply:
There are pros and cons of both: 

1. If instances are accepted in the constructor then consumer needs to write extra code and need to have extra knowledge of contract instances. Which somewhat exposes inner details. 

2. If instances are not accepted in the constructor then it's definitely less code for the consumer. But this makes constructor hard to test. Though other functions are properly testable. 

I had this dilemma while writing this. We can discuss about this. 

@hobofan @benjaminbollen  Thoughts? 

>

Reply:
It's not only about testing, it's also about maintainability and extendability. For example, let's say you now want to run this for a BrandedToken instead of an EIP20Token. If you inject the dependency (and ideally expect an interface), then you can use the same code with the BrandedToken *without changing any existing code in the Staker.* You only need to make sure that both tokens implement the same interface.
On the other hand, if you hardcode EIP20Token everywhere in the project, you would need to change all those places in the entire project to change from EIP20Token to BrandedToken. Or add switches to select one. Then later you want to add a third one ...

This is just one example, but it should make the point clear. It's a software engineering best practice to not hard-code dependencies. It's a classical application of loosely coupled composition.

To your point 1. this means the user could *choose* whether to stake a branded token or an eip20token.

That being said, you can choose your user facing interface carefully and you may want to hide the GatewayComposer. Even though, as this is branded token, I personally would say that it is ok for the user to know about the gateway composer.

Reply:
Now that I think about it, we could consider `Mosaic.ContractInteract.EIP20Token` an *interface* as it is only dependent on the ABI 🤔

Reply:
Created #110  for this.